028-service-binding.md

```diff
@@ -0,0 +1,276 @@
+
+This proposal will introduce a new `KafkaAccess` operator. The operator will be separate from the existing `strimzi-kafka-operator` and live in its own GitHub repo in the `strimzi` organization.
+
+The actions taken during the reconcile loop of the new operator will be:
+1. The operator can be deployed at cluster-scope where it will watch resources across all namespaces, or in a single namespace, where it will only watch resources in that namespace.
+2. On start-up the operator will watch all resources of Kind `KafkaAccess`.
+3. For any `KafkaAccess` resource present it will add a status of NotReady to that CR.
+4. For each `KafkaAccess` resource it will perform a GET request on the Kafka cluster that is 
+referenced in the `KafkaAccess` and pick out the configured listeners.
+   1. If this or any subsequent GET requests fail due to a 401 (unauthorised) error, the status reason will be updated to reflect this.
+5. The operator will determine which listener to choose:
+   1. If there are no listeners in the `Kafka` CR the reconcile loop will fail.
+   2. If the `KafkaAccess` specifies a listener, the operator will choose this listener.
+   3. If there are no listeners specified in the `KafkaAccess`, but only one listener in the `Kafka` CR, the operator will choose that listener.
+   4. If there are no listeners specified in the `KafkaAccess`, but multiple listeners listed in the `Kafka` CR, the operator 
+   will filter the list by comparing the `tls` and `authentication` properties in the `Kafka` and `KafkaUser` CRs.
+   5. If there are still multiple listeners after filtering for `tls` and `authentication` the operator will choose the one that is of type `internal`.
+   6. If there are still multiple listeners after filtering for `internal` the operator will sort the listeners alphabetically by name and choose the first one.
+   7. If the operator fails to find a suitable listener after filtering the reconcile loop will fail.
+8. The operator will then create a secret and add the bootstrapServers for the chosen listener to that secret.
+9. If the chosen listener uses TLS, the operator will then perform a GET request on the `cluster-ca-cert` secret and put the certificate and truststore files in the new secret.
+10. If the `KafkaAccess` references a `KafkaUser`, the operator will perform a GET request on that user to find the name of the secret containing the credentials and the username 
+from the status if it is provided. The operator will also perform a GET request on the user secret to get any credentials. 
+11. The operator will add any user related certificates and credentials to the secret it created earlier.
+12. The operator will add a `status.binding` to the `KafkaAccess` resource with the name of the secret. 
+13. The operator will then use a `Watch` or `Informer` to keep track of changes to the `Kafka` CR and, if used earlier, the `cluster-ca-cert` secret, the `KafkaUser` resource and the secret created by the `KafkaUser` operator,
```

Comment:
I think this goes beyond the proposal and more into the implementation realm. But I think it will not scale well to have a separate watch/informer for each `Kafka` and `KafkaUser`. Maybe it should have a single watch/infromer for these and cross reference the Kafka, KafkaUser and KafkaAccess resources internally.

Reply:
I agree this is more implementation realm, but interestingly this is solved quite nicely I think by java-operator-sdk. It allows you to watch all `Kafka` and `KafkaUser` resources and then trigger the reconcile for the relevant `KafkaAccess` resources when they change.